pkg/collector/corechecks/cluster/ksm/kubernetes_state.go

```diff
@@ -775,17 +778,22 @@ func (k *KSMCheck) getClusterName() {
 // Sets the kube_cluster_name tag for all metrics.
 // Adds the global user-defined tags from the Agent config.
 func (k *KSMCheck) initTags() {
-	if k.instance.Tags == nil {
-		k.instance.Tags = []string{}
-	}
+	k.globalTags = []string{}
 
 	if k.clusterNameTagValue != "" {
-		k.instance.Tags = append(k.instance.Tags, "kube_cluster_name:"+k.clusterNameTagValue)
+		k.globalTags = append(k.globalTags, "kube_cluster_name:"+k.clusterNameTagValue)
 	}
 
 	if !k.instance.DisableGlobalTags {
-		k.instance.Tags = append(k.instance.Tags, configUtils.GetConfiguredTags(k.agentConfig, false)...)
+		k.globalTags = append(k.globalTags, configUtils.GetConfiguredTags(k.agentConfig, false)...)
 	}
+
+	// This function guarantees:
+	// - len(k.globalTags) == cap(k.globalTags) => Any call to append() will copy the underlying array, making append() safe
```

Comment:
I think it is in the Go spec https://go.dev/ref/spec#Appending_and_copying_slices 
> If the capacity of s is not large enough to fit the additional values, append [allocates](https://go.dev/ref/spec#Allocation) a new, sufficiently large underlying array that fits both the existing slice elements and the additional values. Otherwise, append re-uses the underlying array.

> I would rather do an explicit copy(…) call when a copy is needed rather than making assumptions about how future calls to append(…) might behave.
Indeed this might be the only solution. The `RawSender` is not safe, thus we should not reuse the arrays after using them as argument.

Reply:
> > > 	// This function guarantees:
	// - len(k.globalTags) == cap(k.globalTags) => Any call to append() will copy the underlying array, making append() safe

> > Is it something that can be proved formally from the GO language spec and that is guaranteed to remain true in any future version of GO ?
Or is it something observed on a specific version of the GO and that relies on implementation details that might change in any future version of GO ?

> If the capacity of s is not large enough to fit the additional values, append allocates a new, sufficiently large underlying array that fits both the existing slice elements and the additional values. Otherwise, append re-uses the underlying array.

> I think it is in the Go spec https://go.dev/ref/spec#Appending_and_copying_slices

The spec says: `a new, sufficiently large underlying array`. For me, it clearly indicates that the re-allocation strategy is an undocumented implementation detail.
So, I still cannot understand what proves that `len(k.globalTags) == cap(k.globalTags)`.

Reply:
Indeed, it looks more like an implementation detail. I think we can't entirely rely on it and should use a copy whenever possible.

 Moreover I wonder if we can at least take into consideration this detail and make sure that some arrays keep the same length and capacity to avoid accidentally recreating an issue similar to the one in this PR. It shouldn't have a negative impact.